instance/models/instance.py

```diff
@@ -455,29 +461,53 @@ def run_playbook(self):
         """
         Run a playbook against the instance active servers
         """
-        with open_repository(self.ansible_source_repo_url,
-                             ref=self.configuration_version) as configuration_repo:
-            playbook_path = os.path.join(configuration_repo.working_dir, 'playbooks')
-            requirements_path = os.path.join(configuration_repo.working_dir, 'requirements.txt')
-
-            self.logger.info('Running playbook "%s/%s"...', playbook_path, self.ansible_playbook_filename)
-
-            log_lines = []
-            with ansible.run_playbook(
-                requirements_path,
-                self.inventory_str,
-                self.vars_str,
-                playbook_path,
-                self.ansible_playbook_filename,
-                username=settings.OPENSTACK_SANDBOX_SSH_USERNAME,
-            ) as processus:
-                for line in processus.stdout:
-                    line = line.decode('utf-8').rstrip()
-                    self.logger.info(line)
-                    log_lines.append([line.rstrip()])
-
-        self.logger.info('Playbook run completed')
-        return log_lines
+        attempt = 0
+        while attempt < self.attempts:
+            with open_repository(self.ansible_source_repo_url,
+                                 ref=self.configuration_version) as configuration_repo:
+                playbook_path = os.path.join(configuration_repo.working_dir,
+                                             'playbooks')
+                requirements_path = os.path.join(configuration_repo.working_dir,
+                                                 'requirements.txt')
+
+                log = ('Running playbook "{path}/{name}" for instance {instance}, '
+                       'attempt {attempt} of {attempts}:').format(
+                           path=playbook_path,
+                           name=self.ansible_playbook_name,
+                           instance=self,
+                           attempts=self.attempts,
+                           attempt=attempt + 1)
+                self.logger.info(log)
+
+                log_lines = []
+                with ansible.run_playbook(
+                    requirements_path,
+                    self.inventory_str,
+                    self.vars_str,
+                    playbook_path,
+                    self.ansible_playbook_filename,
+                    username=settings.OPENSTACK_SANDBOX_SSH_USERNAME,
+                ) as processus:
+                    output = itertools.chain(
+                        zip(itertools.repeat(logging.INFO), processus.stdout),
+                        zip(itertools.repeat(logging.ERROR), processus.stderr)
+                    )
```

Comment:
@chromano One more thought on this issue.  All I said so far was assuming that we actually need to read from two different streams.  I'm not sure we really need to distinguish between the two streams in this case.  We could simply set `stderr=subprocess.STDOUT` in the `Popen()` call to run the playbook and would only have to read from a single stream.  We wouldn't be able to set the log levels as you did anymore, but it would save us the trobule mentioned earlier.